### PR TITLE
Optimize manyN to not require backtracking

### PR DESCRIPTION
**Pull request by jinahong**
Hello, I would like to suggest a new implementation for the `manyN` function, which consumes a parser exactly N times and does not require any backtracking.
The current implementation of the `manyN` function uses the `takeTillWithCount` function which performs N+1 parsing attempts followed by a backtracking. It seems better not to perform unnecessary computations.

### CODE REVIEW

**Comment by devshorts** [comment]
Thanks for the submission! I'll take a look at this this week. Thanks for contributing

File: Combinator/Combinator.fs [hunk 1]
```diff
@@ -167,14 +167,26 @@ module Combinator =
...
+            | x when x > 0 -> Some (List.rev found), currentState
+            | _ -> None, currentState
+        let rec many' parser (resultList, currentState: State<_,_,_>, cnt) =
+            let returnValue acc = didFind acc currentState
+            match currentState.hasMore() with
+            | false -> returnValue resultList
+            | true ->
+                match parser currentState with
+                | Some (result), nextState ->
+                    if cnt >= num then returnValue (result :: resultList)
```
Comment (devshorts) [inline Combinator/Combinator.fs]: Should this be equal to and not just greater than or equal to?  We want to not accidentally parse too many items, it specifically needs to be N items or fail

**Comment by devshorts** [comment]
Can you do a rebase from master so that this PR runs against the travis unit tests? Thanks

**Comment by jinahong** [comment]
The previous PR failed unittest, so I modified the PR to handle corner cases in the same way as yours.

File: Combinator/Combinator.fs (hunk 1 above)
Comment (jinahong) [inline Combinator/Combinator.fs]: good point, Thanks!

File: Combinator/Combinator.fs [hunk 2]
```diff
@@ -169,14 +169,26 @@ module Combinator =
...
+        let rec many' parser (resultList, currentState: State<_,_,_>, cnt) =
+            match currentState.hasMore() with
+            | false -> failwith ("Error, There is nothing to attempt to match")
+            | true ->
+                match parser currentState with
+                | Some result, nextState ->
+                    if cnt = num then didFind (result :: resultList) nextState
+                    elif cnt < num then many' parser (result :: resultList, nextState, cnt + 1)
+                    else Some [], currentState
+                | _ -> failwith ("Needed to consume at least " + string num + " element but did not")
```
Comment (devshorts) [inline Combinator/Combinator.fs]: Can we make this say something like "Needed to consume N items but only found Y items"? This gives a better indication of how many were found vs expected

File: Combinator/Combinator.fs [hunk 3]
```diff
@@ -169,14 +169,26 @@ module Combinator =
...
+        many' parser ([], state, 1)
```
Comment (devshorts) [inline Combinator/Combinator.fs]: I'm not sure I like seeding this as 1 based.  This kind of implies that we already found an item right? I think this should be seeded as 0, indicating we haven't found anything yet, and then updating the guard clauses to be 0 based instead of 1 based

File: Combinator/Combinator.fs [hunk 4]
```diff
@@ -169,14 +169,26 @@ module Combinator =
...
-        takeTillWithCount 0 (fun i -> List.length i > num) parser    >>= fun result -> 
-            if result.Length <> num then
-                failwith ("Error, attempted to match " + num.ToString() + " but only got " + result.Length.ToString())
 
-            preturn result          
-    
+
+
+    let manyN num parser state =
+        let didFind found currentState =
```
Comment (devshorts) [inline Combinator/Combinator.fs]: Is this function necessary? It just returns the list reversed and returns the current state.  Seems like we can update the final clause to just do that inline and remove the extra code

File: Combinator/Combinator.fs [hunk 5]
```diff
@@ -169,14 +169,26 @@ module Combinator =
...
-    let manyN num (parser) =                              
-        takeTillWithCount 0 (fun i -> List.length i > num) parser    >>= fun result -> 
-            if result.Length <> num then
-                failwith ("Error, attempted to match " + num.ToString() + " but only got " + result.Length.ToString())
 
-            preturn result          
-    
+
+
+    let manyN num parser state =
```
Comment (devshorts) [inline Combinator/Combinator.fs]: Do we need to include the state argument? It should infer that it takes a state argument already right?  Its not the end of the world to include it but it doesn't follow the pattern that the other partial functions at the top level follow

**Comment by devshorts** [comment]
Are there any other locations in the project that would benefit from an exactly N parser usage? If there are we may want to replace them to avoid the one extra backtracking. Can you take a look and see if there are others? Thanks!

File: Combinator/Combinator.fs (hunk 5 above)
Comment (jinahong) [inline Combinator/Combinator.fs]: Seems like other partial functions, including manyN, are also not necessarily closures. So, I included the state argument. Since a closure is treated as an object, it is better to eliminate needless closures for performance reasons.

File: Combinator/Combinator.fs (hunk 4 above)
Comment (jinahong) [inline Combinator/Combinator.fs]: Seems like other partial functions, including manyN, are also not necessarily closures. So, I included the state argument. Since a closure is treated as an object, it is better to eliminate needless closures for performance reasons.

File: Combinator/Combinator.fs (hunk 3 above)
Comment (jinahong) [inline Combinator/Combinator.fs]: My bad. I totally agree with you.

**Comment by devshorts** [comment]
Cool, sounds good. Give me a few days to give it a final once over and we'll get it in! Thank you for contributing!

**Comment by jinahong** [comment]
VS configuration has automatically removed whitespace... I divided the previous commit into two commits. Thanks for letting me know.

**Comment by devshorts** [comment]
Looks good! Appreciate splitting the commits, it will make the history easier to follow.  I'll merge this and try a release this weekend or next week
